add uncorrected frame to fragment metadata

ibsafa opened the pull request:
### Description

This PR introduces an extra variable in the NTB and TPC fragment metadata. The variables contain the uncorrected frame number (See related PR in sbndaq_artdaq which introduces a correction to the frame). This PR ensures we retain the information about the frame before correction. 

### Related Repository Branches

sbndaq_artdaq: bug/isafa_NTB_Frame_Rollover

### Testing details
_(Note: edit as appropriate.)_
- *Where it was tested*: SBN-ND
- *Run number associated with test*:  
- Other information:

tomjunk commented:
Three comments on this PR:

1.  It changes the data format out of the DAQ.  ROOT provides for schema evolution of data products so that old-format data can be read into programs expecting new-format data.  But the underlying data product is the artdaq::Fragment, and the metadata and payload are defined by the users and not managed as class members that ROOT can adjust.  Thus we need to do the updating of old data ourselves.  And there must be a way to detect whether data are in the old format vs the new format.  One solution is simply not to try to read the old NTB fragments and just use this new format from now on.  Another is to put a version number in the metadata.  Or possibly use the size of the metadata.

2.  The PR re-introduces the bug that was fixed in PR 109:

https://github.com/SBNSoftware/sbndaq-artdaq-core/pull/109

3.  This needs to be propagated to the offline branch too.   I saw that the offline release did not correspond to what was in the tags, but rather what was in the offline branch.  Is this still the case?

tomjunk commented:
Oh, regarding point number 1 above, both the NTB fragment and the TPC fragment have changed size, so this is a breaking change.  If we can query the metadata size we might be able to do the conversion from old data to new data in both the NTB and the TPC decoders.

tomjunk commented:
Converting to draft so it doesn't get merged by accident.  Point number 2 above needs to be fixed before merging -- point number 1 can be worked around by consumers of this data, and point number 3 is just a comment about a need to propagate this to the offline branch.

I poked around the artdaq::Fragment header and the decoder code -- the workaround for the data format change looks less scary than I originally thought.  The pointer to the data payload in the fragment depends on the size of the metadata the fragment was written with, not the new size (which isn't even known when the data are an artdaq::Fragment and not yet a sbndaq::NevisTPCFragment).  So the decoders should pick up the right pointer to the start of the data in both the old and the new formats without having to modify their source code.

That said, the metadata now are different size, and one can get a pointer to the metadata using the metadata() method of NevisTPCFragment.  One would have to query the metadata size (it's in the fragment header) before retrieving the new value, but that may be all that's necessary.

ibsafa commented:
Hi Tom, 

Thanks for taking a look! I addressed point 2 and recommitted the fixed you had introduced in #108 

1. I agree, like you said I believe this is not a breaking change and we added the new value in the metadata at the end for that purpose, so that it's available but doesn't break old decoders. That said, we will still update the decoder to read in the new value, but that does not have to happen before this is merged. 
3. For the offline branch, I will put out another release that points to the new artdaq-core once that is out.

tomjunk reviewed (approved):
Hi Ibrahim -- thanks so much!  This looks good.  I think we're all on the same page!
